CHANGELOG.md

```diff
@@ -1,5 +1,84 @@
 # Changelog
 
+## 0.4.0
+
+### A new, hand-written parser
+
+This release replaces the generated parser with a hand-written recursive descent parser in Ruff.
+There's a lot to say about this exciting change, so check out the [blog post](TODO-LINK) for more details!
+
+See [#10036](https://github.com/astral-sh/ruff/pull/10036) for implementation details.
+
+### A new language server in Rust
+
+With this release, we also want to highlight our new language server. `ruff server` is a Rust-powered language
+server that comes built-in with with Ruff. It can be used with any editor that supports the [Language Server Protocol](https://microsoft.github.io/language-server-protocol/) (LSP).
```

Comment:
found another `with` bug

```suggestion
server that comes built-in with Ruff. It can be used with any editor that supports the [Language Server Protocol](https://microsoft.github.io/language-server-protocol/) (LSP).
```